src/compiler/preprocess/index.ts

```diff
@@ -81,55 +60,124 @@ export default async function preprocess(
 	const markup = preprocessors.map(p => p.markup).filter(Boolean);
 	const script = preprocessors.map(p => p.script).filter(Boolean);
 	const style = preprocessors.map(p => p.style).filter(Boolean);
+	const source_maps: Array<Processed['map']> = [];
+
+	let source_locator: ReturnType<typeof getLocator>;
+
+	function get_replacement(offset: number, original: string, processed: Processed, prefix: string, suffix: string): GeneratedStringWithMap {
+		const generated_prefix = GeneratedStringWithMap.from_source(filename, prefix, source_locator(offset));
+		const generated_suffix = GeneratedStringWithMap.from_source(filename, suffix, source_locator(offset + prefix.length + original.length));
+
+		let generated;
+		if (processed.map) {
+			const full_map = typeof processed.map === "string" ? JSON.parse(processed.map) : processed.map;
+			const decoded_map = { ...full_map, mappings: decode(full_map.mappings) };
+			const processed_offset = source_locator(offset + prefix.length);
+			generated = GeneratedStringWithMap.from_generated(processed.code, offset_source_location(processed_offset, decoded_map));
+		} else {
+			generated = GeneratedStringWithMap.from_generated(processed.code);
+		}
+		const map = generated_prefix.concat(generated).concat(generated_suffix);
+		return map;
+	}
+
+	async function replace_async(str: string, re: RegExp, func: (...any) => Promise<GeneratedStringWithMap>): Promise<GeneratedStringWithMap> {
+		const replacement_promises: Array<Promise<Replacement>> = [];
+		str.replace(re, (...args) => {
+			replacement_promises.push(
+				func(...args).then(
+					(replacement) =>
+						({
+							offset: args[args.length - 2],
+							length: args[0].length,
+							replacement
+						}) as Replacement
+				)
+			);
+			return '';
+		});
+		const replacements = await Promise.all(replacement_promises);
 
+		let out: GeneratedStringWithMap;
+		let last_end = 0;
+		for (const { offset, length, replacement } of replacements)
+		{
```

Comment:
I think this should probably be on the previous line for consistency with the existing code